### PR TITLE
add eos_pyateos module

### PR DESCRIPTION
**Pull request by lvrfrc87**
##### SUMMARY
This is a new module for `eos` operational tests. Based on [pyATEOS](https://gitlab.com/networkAutomation/pyateos#pyateos) framework. 
From README.md: 

> A snapshot of the operational status of an Arista switch is taken before a config or network change and compare against a second snapshot taken after the change. A diff file is generated in .json format.

I understand that from Ansible 2.10 the modules would be in collections but would be really nice and useful if we can have this merged in 2.9 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
eos_pyateos

##### ADDITIONAL INFORMATION
Playbook example:
```
- name: run BEFORE tests.
  eos_pyateos:
      before: true
      test:
          - acl
      group: 
          - mgmt
          - layer2
      hostname: "{{ inventory_hostname }}"
  register: result

- name: save BEFORE file IDs.
  delegate_to: 127.0.0.1
  set_fact:
      before_ids: "{{ result.before_file_ids }}"

- name: change mgmt config on switch.
  eos_config:
      lines:
          - no ntp server vrf mgmt 10.75.33.5
          - ntp server vrf mgmt 216.239.35.4
          - no snmp-server host 10.1.22.1 vrf mgmt version 2c snmp_pass
          - snmp-server host 10.1.22.9 vrf mgmt version 2c snmp_pass

- name: shutdown interface.
  eos_config:
      lines:
          - shutdown
      parents: interface Ethernet50/1

- name: edit ACL.
  eos_config:
      lines:
          - no 10
          - 10 remark pyATEOS TEST
      parents: ip access-list standard SNMP

- name: run AFTER tests.
  eos_pyateos:
      after: true
      test:
          - acl
      group:
          - mgmt
          - layer2
      hostname: "{{ inventory_hostname }}"
  register: result

- name: save AFTER file IDs.
  delegate_to: 127.0.0.1
      set_fact:
          after_ids: "{{ result.after_file_ids }}"
 
- name: run DIFF result.
  eos_pyateos:
      compare: true
      group:
          - mgmt
          - layer2
      test:
          - acl
      hostname: "{{ inventory_hostname }}"
      filter: true
      files: 
          - "{{ before_ids }}"
          - "{{ after_ids }}"
```

### CODE REVIEW

**Comment by pabelanger** [comment]
Thanks!

To fix the lint errors, you'll want to run
```
tox -eblack
```
This will auto-format the code to help keep flake8 and style-guide happy. We'll also want to add some integration tests to validate this code path works. You can look into tests/integration folder for an example.

**Comment by FedericoOlivieri** [comment]
Hello!
I have added some integration tests. Not having access to the switch I am not 100% sure if all commands available in pyateos are supported so this is the reason why I have chosen the most common tests. Please, let me know if that is ok.

**Comment by FedericoOlivieri** [comment]
@pabelanger I can see integration test failed but I cannot find the reason why. Can you please help me with that?

Thanks!

**Comment by pabelanger** [comment]
It looks like your assert statements are not formatted properly:

  https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_9/9/3aa20625fe9e007dcc969e5142470d20b1401ddf/check/ansible-test-network-integration-eos-python27/788efde/controller/ara-report/result/a754f5ad-90fd-4455-b4e1-8907c30d6adb/

**Comment by FedericoOlivieri** [comment]
@pabelanger thanks for that. I had a look around to Zuul outputs and I could not find anything useful apart long debugs. For my knowledge, so in future I won't need to bother you: where can I find the output you sent me?

**Comment by pabelanger** [comment]
The 'checks' tab at the top of the page will leave results like:

  https://dashboard.zuul.ansible.com/t/ansible/build/788efde485f247959aa1bad59deea41c

you can that look at 'log url' link,  this gets you to https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_9/9/3aa20625fe9e007dcc969e5142470d20b1401ddf/check/ansible-test-network-integration-eos-python27/788efde/

If you then follow controller folder, you will see all the debug related, ara logs we collect from controller side.  If looking at ARA, look for the red X, then keep expanding until you see the failure.

Otherwise, you can inspect the ansible-debug.txt files

**Comment by FedericoOlivieri** [comment]
@pabelanger I get this error from integration test `msg": "Could not find imported module support code for arista.eos.eos_pyateos.  Looked for either run_commands.py or eos.py"`

I could not find any useful info around. Any idea why this is happening?

File: plugins/modules/eos_pyateos.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+"""
+import os
+import re
+import sys
+import json
+import time
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.network.eos.eos import run_commands
+from ansible.module_utils.network.eos.eos import eos_argument_spec
+from ansible.module_utils.network.eos.eos import get_connection
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
from ansible_collections.arista.eos.plugins.module_utils.network.eos.eos import run_commands
from ansible_collections.arista.eos.plugins.module_utils.network.eos.eos import eos_argument_spec
from ansible_collections.arista.eos.plugins.module_utils.network.eos.eos import get_connection
```

File: plugins/modules/eos_pyateos.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "network",
+}
+
+DOCUMENTATION = """
+---
+module: eos_pyateos
+version_added: "2.9"
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
version_added: "2.10"
```

File: plugins/modules/eos_pyateos.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+        argument_spec=argument_spec,
+        mutually_exclusive=mutually_exclusive,
+        required_if=required_if,
+        required_one_of=required_one_of,
+    )
+
+    if not HAS_JSONDIFF:
+        return module.fail_json(
+            msg="jsondiff is not installed, try 'pip install jsondiff'"
+        )
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
        return module.fail_json(msg=missing_required_lib("jsondiff"))
```

File: plugins/modules/eos_pyateos.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+
+    if not HAS_JMESPATH:
+        return module.fail_json(
+            msg="jmespath is not installed, try 'pip install jmespath'"
+        )
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
        return module.fail_json(msg=missing_required_lib("jmespath"))
```

File: plugins/modules/eos_pyateos.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+        }
+    }
+}
+"""
+import os
+import re
+import sys
+import json
+import time
+from ansible.module_utils.basic import AnsibleModule
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
from ansible.module_utils.basic import AnsibleModule, missing_required_lib
```

File: plugins/modules/eos_pyateos.py (hunk 1 above)
Comment (Qalthos) [reply plugins/modules/eos_pyateos.py]: There might be more things, but this is just the easy picks from the errors shown

File: plugins/modules/eos_pyateos.py (hunk 5 above)
Comment (Qalthos) [reply plugins/modules/eos_pyateos.py]: This generates a standard message for missing libraries so they always look the same and you don't have to reinvent it every time.

**Comment by FedericoOlivieri** [comment]
@Qalthos thanks for you valuable suggestions :). I have added all the changes suggested.

File: plugins/modules/eos_pyateos.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+        }
+    }
+}
+"""
+import os
+import re
+import sys
+import json
+import time
+from ansible_collections.arista.eos.plugins.module_utils.basic import AnsibleModule, missing_required_lib
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]: `basic` is still part of ansible, and not this collection (:

```suggestion
from ansible.module_utils.basic import AnsibleModule, missing_required_lib
```

File: tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      - mgmt
+    test:
+      - prefix_list
+      - vlan
+      - arp
+    hostname: "{{ inventory_hostname }}"
+    provider: '{{ eapi }}'
+  register: result
+- assert:
+    that: (result.before_file_ids|lenght=5)
```
Comment (Qalthos) [inline tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml]:
```suggestion
    that: (result.before_file_ids|length == 5)
```

File: tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    test:
+      - prefix_list
+      - vlan
+      - arp
+    hostname: "{{ inventory_hostname }}"
+    provider: '{{ eapi }}'
+  register: result
+- assert:
+    that:
+      - (result.after_file_ids|lenght=5)
```
Comment (Qalthos) [inline tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml]:
```suggestion
      - (result.after_file_ids|length == 5)
```

File: plugins/modules/eos_pyateos.py [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+module: eos_pyateos
+version_added: "2.10"
+author:
+  - "Federico Olivieri (@Federico87)"
+
+short_description: Operational status tests on Arista device.
+description: A snapshot of the operational status of a switch is taken before a
+    config or network change and compare against a second snapshot taken after the change.
+    A diff file is generated in .json format.
+extends_documentation_fragment: eos
```
Comment (Qalthos) [inline plugins/modules/eos_pyateos.py]:
```suggestion
```

All this does is define the `provider` argument (and by extension `connection: local`), which we are not supporting for new modules. You should also likewise remove `provider: '{{ eapi }}'` from your tests below, as even if you were intending to support it, there is no actual testing using `connection: local`

Comment (FedericoOlivieri) [reply plugins/modules/eos_pyateos.py]: Ok. I just reverse engineering a test module from collection. I was not 100% sure how the session was handled so I decided to keep like the example I was following

File: tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml (hunk 7 above)
Comment (FedericoOlivieri) [reply tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml]: Sure. I need to set-up the environment. Then I ll make sure to run the test locally first

**Comment by FedericoOlivieri** [comment]
Test failed because missing requirements for `jsondiff` and `jmespath`. I have tried to install them via playbook with `delegte_to: 127.0.0.1` but I get the below. Is Ansible running within EOS?

```
2020-03-22 10:28:55.216887 | fedora-31 | TASK [eos_pyateos : install python3-pip] ***************************************
2020-03-22 10:28:55.217129 | fedora-31 | task path: /home/zuul/.ansible/collections/ansible_collections/arista/eos/tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml:4
2020-03-22 10:28:55.278206 | fedora-31 | fatal: [eos-4.20.10]: FAILED! => {
2020-03-22 10:28:55.278253 | fedora-31 |     "msg": "Unable to automatically determine host network os. Please manually configure ansible_network_os value for this host"
```

Here the playbook:

```
- name: install python3-pip
  delegate_to: 127.0.0.1
  yum:
    name: python3-pip
    state: present
  become: yes

- name: install jsondiff,jmespath
  delegate_to: 127.0.0.1
  pip:
    name: "{{ item }}"
  loop:
    - jsondiff
    - jmespath
```

**Comment by FedericoOlivieri** [comment]
@Qalthos @pabelanger could you please have a look to the comment above? I cannot install requirement on test env.
Thanks!!!!

**Comment by pabelanger** [comment]
You can add them into https://github.com/ansible-collections/eos/blob/master/requirements.txt, if the new module needs more python dependencies

File: tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  delegate_to: 127.0.0.1
+
+
+- name: install jsondiff,jmespath
+  delegate_to: 127.0.0.1
+  pip:
+    name: "{{ item }}"
+  loop:
+    - jsondiff
+    - jmespath
```
Comment (pabelanger) [inline tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml]: This isn't needed, you can add them into https://github.com/ansible-collections/eos/blob/master/requirements.txt

Comment (FedericoOlivieri) [reply tests/integration/targets/eos_pyateos/tests/eapi/pyateos.yaml]: Thanks! I am still trying to get my way around the new set-up and tests :)

**Comment by justjais** [comment]
recheck

**Comment by FedericoOlivieri** [comment]
@justjais what check stays for? Trigger CI/CD via push? Thanks :)

**Comment by FedericoOlivieri** [comment]
@pabelanger  I get this error in CI but not in my local test env `msg": "Unable to automatically determine host network os. Please manually configure ansible_network_os value for this host"`

**Comment by justjais** [comment]
@FedericoOlivieri The tests are failing coz of following error: 
https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_9/9/cc1f04ccf5ff3f3f3d1c1689bbd8342365b8dab7/check/ansible-test-network-integration-eos-python27/af62ca9/controller/ara-report/reports/2c545478-e55c-4716-b063-c1c45b6a0ee3.html

Please fix the same.

**Comment by lvrfrc87** [comment]
@justjais @Qalthos would be possible to have code review for this PR please?
So, if I have to make some changes I will do them while I make the module compatible with python 2.7

**Comment by abenokraitis** [comment]
This module should be in its own collection, outside of `arista.eos` so that the pyateos community is not handcuffed by the Ansible team for inclusion and maintenance criteria. I would say this could be part of `community.network` or part of its own `arista.pyateos` collection.

**Comment by lvrfrc87** [comment]
@abenokraitis thanks for your feedback. So, do you want me to close this PR and raise a new one for `community.network`. In case I believe module name has to be changed from `eos_pyateos` in something like `pyateos_ansible` ?

**Comment by abenokraitis** [comment]
> @abenokraitis thanks for your feedback. So, do you want me to close this PR and raise a new one for `community.network`. In case I believe module name has to be changed from `eos_pyateos` in something like `pyateos_ansible` ?

Yup, I think this makes total sense for where. As for the naming, I'd probably just keep the same module name given it's not in its own collection. I'd maybe ask in the Network Community weekly meeting though.

**Pull request closed without merging**